Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs

```diff
@@ -1798,7 +1798,13 @@ private void TryPrepareFileAccessMonitoringForPip(Process pip, HashSet<AbsoluteP
                             FileAccessPolicy.AllowAll | // Symlink creation is allowed under opaques.
                             FileAccessPolicy.AllowRealInputTimestamps |
                             // For shared opaques, we need to know the (write) accesses that occurred, since we determine file ownership based on that.
-                            (directory.IsSharedOpaque? FileAccessPolicy.ReportAccess : FileAccessPolicy.Deny);
+                            (directory.IsSharedOpaque? FileAccessPolicy.ReportAccess : FileAccessPolicy.Deny) |
+                            // For shared opaques and if allowed undeclared source reads is enabled, make sure that any file used as an undeclared input under the 
+                            // shared opaque gets deny write access. Observe that with exclusive opaques they are wiped out before the pip runs, so it is moot to check for inputs
+                            // TODO: considering configuring this policy for all shared opaques, and not only when AllowedUndeclaredSourceReads is set. The case of a write on an undeclared

```

Comment:
>The case of a write [](start = 152, length = 19)

Can you elaborate more (in this comment) why the case of write on undeclared input is more likely to happen when undeclared sources are allowed?

Reply:
Because the mode implies that many legit source files are not declared, so the chances of writing into an existing file that is not caught by the regular manifest is higher.

There is another secondary reason as well, which I didn't think it was worth writing down: by doing it in this way we are technically enabling this option for MSBuild/Ninja only. After giving this some mileage, we could consider extending this for all shared opaques.

---
In reply to: [299183232](https://github.com/microsoft/BuildXL/pull/536#discussion_r299183232) [](ancestors = 299183232)